### ISSUE TITLE
Review permission handling in entityprovider

### ISSUE THREAD

**Issue by steveswinsburg** [sakaiproject/sakai#1366]
_From @steveswinsburg on July 23, 2015 22:16_

Getting this in the logs:

````
INFO Could not process entity: /gbng (403)[null]: Security exception handling request for view (/gbng.json), this is typically caused by the current user not having access to the data requested or the user not being logged in at all :: message=You do not have instructor-type permissions in this site. (rethrown)
````
Need to review how the permissions are checked in the entity provider, and allow TA to access some of the calls.

_Copied from original issue: steveswinsburg/gradebookNG#154_

**Comment by payten** [comment]
Hi @kyleblythe, @steveswinsburg.  I think I found the source of the `403` error in the logs.  The AJAX request to check for `isotheruserediting` throws a SecurityException when the user isn't an instructor.

Reviewing the service, here are the actions that require some kind of instructor or TA check:

- **/direct/gbng/assignments** (`getAssignmentList`): I can't seem to find any uses of this endpoint? @steveswinsburg, perhaps we don't need this anymore?
- **/direct/gbng/assignment-order** (`updateAssignmentOrder`): used by the custom drag-and-drop javascript to save the uncategorised sort order changes.  Should the TA be able to reorder columns at all?
- **/direct/gbng/isotheruserediting** (`isAnotherUserEditing`): used by the custom javascript to check for and list changes from other users on the current gradebook and node. It makes sense for TAs to have access to this endpoint.
- **/direct/gbng/categorized-assignment-order** (`updateCategorizedAssignmentOrder`): used by the custom drag-and-drop javascript to save categorised sort order changes. As above, should the TA be able to reorder columns?

My feeling is that:

1. TA's should be able to reorder columns, so we disable drag and drop for them and also retain the instructor-check on those endpoints.
2. We remove the `getAssignmentList` endpoint if it's not being used.
3. And finally, we allow TA's to view other user edits. 

Does that seem ok with you guys?

**Comment by kyleblythe** [comment]
agreed 100% re: the ability to reorder items, @payten . I also like that the TA cannot create GB items or edit GB items, since they're just graders (or viewers). But I also want to get @jeffpasch 's input on that as well.

I am noticing a few things, though:

* the column header dropdown appears to be missing entirely for a TA with grade permissions. Given that they have the ability to see cell-by-cell grade logs, I would say that they should be able to view grade statistics and set score for empty cells. 

This relates to an enhancement that we talked about a while back and noted in #1376 , wherein there would be section awareness for the set score for empty cells feature. This would allow instructors to apply a default score to empty cells for the entire GB or just specific sections. For TA's however, it would allow them to set a default score for only the section(s) for which they have grading access.

* The TA currently has the ability to set a course grade override (even when the course grade isn't visible to them). They should not be able to access this option when the course grade permission checkbox is not selected for them.

On a related note, should they have the ability to provide a course grade override at all? Currently, the permission checkbox is stating that it allows the TA to "View" the course grade.

* Lastly, right now the TA has access to the Course Grade Display dropdown option in the Course Grade header, as well as the GB-wide set 0 for empty cells option. They likely should be disabled for TA's as well.

**Comment by payten** [comment]
In implementing the menu behaviour above, I found a few new TA-related bugs: #1832, #1839 

Also, set score for empty cells still relies on a solution to the open issue #1376.

**Comment by kyleblythe** [comment]
Hey @steveswinsburg and @payten ,

This is all looking great! One thing I'm seeing, that I believe we had documented in another issue but that I cannot find at the moment re: TA permissions.

Currently, the TA's section/group filter dropdown shows as "All Sections/Groups," regardless of which section(s) the TA has access to:

![image](https://cloud.githubusercontent.com/assets/11447474/14287653/845901e0-fb21-11e5-84d6-5370bb8f238d.png)

I think we had discussed the following:

If the TA does **not** have grade or view access to all sections/groups, this filter should reflect the sections/groups to which they **do** have access. So, for example, if a GB has 4 sections, and the TA has access to grade section 001 and view section 002, then the filter should default to "Available Sections/Groups" with the dropdown options showing the sections/groups. I think that'll clarify this a bit.

And, if the TA has grade/view access for only one section/group, I don't think there needs to be a dropdown at all. It can just say "Section 001 - Showing 7 Students". Make sense?
